File: 06_frequency_each_char.py
"""
Task 6: Find the frequency of each character in a string.
Prints each character and its count in order of first occurrence.
"""
import sys
from collections import OrderedDict

def freq_each_char(s: str):
    d = OrderedDict()
    for ch in s:
        d[ch] = d.get(ch, 0) + 1
    return d

if __name__ == '__main__':
    s = sys.stdin.readline().rstrip('\n')
    d = freq_each_char(s)
    for ch, cnt in d.items():
        print(f"{ch}: {cnt}")

# Time Complexity: O(n)
# Space Complexity: O(k)

File: 06_frequency_each_char.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Task 6: Find the frequency of each character in a string.
Prints each character and its count in order of first occurrence.
*/

int main()
{
    string s;
    if (!getline(cin, s))
        return 0;
    unordered_map<char, int> cnt;
    vector<char> order;
    for (char c : s)
    {
        if (cnt[c]++ == 0)
            order.push_back(c);
    }
    for (char c : order) cout << c << ": " << cnt[c] << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(k) where k is number of distinct chars
